lib/c-api/src/wasm_c_api/module.rs

```diff
@@ -115,3 +157,376 @@ pub unsafe extern "C" fn wasm_module_serialize(
+
+                    const wasm_limits_t* table_limits = wasm_tabletype_limits(table_type);
+                    assert(table_limits->min == 1);
+                    assert(table_limits->max == 2);
+
+                    wasm_externtype_delete((wasm_externtype_t*) extern_type);
+                }
+
+                {
+                    const wasm_importtype_t* import_type = import_types.data[3];
+
+                    const wasm_name_t* import_module = wasm_importtype_module(import_type);
+                    wasmer_assert_name(import_module, "ns");
+
+                    const wasm_name_t* import_name = wasm_importtype_name(import_type);
+                    wasmer_assert_name(import_name, "memory");
+
+                    const wasm_externtype_t* extern_type = wasm_importtype_type(import_type);
+                    assert(wasm_externtype_kind(extern_type) == WASM_EXTERN_MEMORY);
+
+                    wasm_memorytype_t* memory_type = wasm_externtype_as_memorytype((wasm_externtype_t*) extern_type);
+                    const wasm_limits_t* memory_limits = wasm_memorytype_limits(memory_type);
+                    assert(memory_limits->min == 3);
+                    assert(memory_limits->max == 4);
+
+                    wasm_externtype_delete((wasm_externtype_t*) extern_type);
+                }
+
+                wasm_importtype_vec_delete(&import_types);
+                wasm_module_delete(module);
+                wasm_byte_vec_delete(wasm);
+                wasm_store_delete(store);
+                wasm_engine_delete(engine);
+            }
+        })
+        .success();
+    }
+
+    #[test]
+    fn test_module_serialize() {
+        (assert_c! {
+            #include "tests/wasmer_wasm.h"
+
+            int main() {
+                wasm_engine_t* engine = wasm_engine_new();
+                wasm_store_t* store = wasm_store_new(engine);
+
+                wasm_byte_vec_t wat;
+                wasmer_byte_vec_new_from_string(&wat, "(module)");
+                wasm_byte_vec_t* wasm = wat2wasm(&wat);
+
+                wasm_module_t* module = wasm_module_new(store, wasm);
+                assert(module);
+
+                wasm_byte_vec_t serialized_module;
+                wasm_module_serialize(module, &serialized_module);
+                assert(serialized_module.size > 0);
+
+                wasm_module_delete(module);
+                wasm_byte_vec_delete(&serialized_module);
+                wasm_byte_vec_delete(wasm);
+                wasm_store_delete(store);
+                wasm_engine_delete(engine);
+            }
+        })
+        .success();
+    }
+
+    #[test]
+    fn test_module_serialize_and_deserialize() {
+        (assert_c! {
+            #include "tests/wasmer_wasm.h"
+
+            int main() {
+                wasm_engine_t* engine = wasm_engine_new();
+                wasm_store_t* store = wasm_store_new(engine);
+
+                wasm_byte_vec_t wat;
```

Comment:
This isn't deleted.

I see that `lib/c-api/tests/wasm_c_api/test-wat2wasm.c` doesn't delete it either, but I can't see how that isn't just a leak? If `wat2wasm` takes ownership, then that fact needs to be added to the function comment in `wasmer_wasm.h`.

(A tip, any program you've built can be run with `valgrind program` to see leaks on exit.)

Reply:
`wat2wasm` must not take ownership of the bytes. I'm fixing that. Thanks!